File: apps/web/tests/contracts/freeLimits.contract.test.js
// Contract tests for the free-tier caps.
//
// These assert the shared fixtures in packages/contracts, which the Dart
// implementation in apps/mobile is held to as well. The caps have to match
// exactly: mobile enforces the trade window by tombstoning rows, so a web app
// that kept more would watch mobile delete trades it had saved.
import contract from '../../../../packages/contracts/free_limits.json';
import {
    FreeLimits,
    canAddDistinctCard,
    cardsFor,
    tradesOverFreeLimit,
} from '../../src/utils/freeLimits.js';

describe('free limits contract', () => {
    it('agrees with the shared caps', () => {
        expect(FreeLimits).toEqual(contract.limits);
    });

    contract.cases
        .filter((testCase) => testCase.limit === 'savedTrades')
        .forEach((testCase) => {
            it(testCase.name, () => {
                // Saving is never refused, only trimmed. Anything else would lose the
                // trade the customer just built.
                expect(testCase.allowed).toBe(true);
                expect(tradesOverFreeLimit(testCase.existing + 1)).toBe(testCase.trimmed);
            });
        });

    contract.cases
        .filter((testCase) => testCase.limit === 'binderCards' || testCase.limit === 'wantListCards')
        .forEach((testCase) => {
            it(testCase.name, () => {
                const isWanted = testCase.limit === 'wantListCards';
                expect(canAddDistinctCard(testCase.existing, { isWanted, isPro: false })).toBe(
                    testCase.allowed,
                );
            });
        });

    it('maps cardsFor to the binder vs want caps', () => {
        expect(cardsFor({ isWanted: false })).toBe(FreeLimits.binderCards);
        expect(cardsFor({ isWanted: true })).toBe(FreeLimits.wantListCards);
    });

    it('never blocks Pro binder or want-list adds', () => {
        expect(canAddDistinctCard(100, { isWanted: false, isPro: true })).toBe(true);
        expect(canAddDistinctCard(100, { isWanted: true, isPro: true })).toBe(true);
    });

    it('trims nothing for an account inside the window', () => {
        expect(tradesOverFreeLimit(0)).toBe(0);
        expect(tradesOverFreeLimit(FreeLimits.savedTrades)).toBe(0);
    });
});

File: apps/mobile/test/contracts/free_limits_contract_test.dart
// Contract tests for the free-tier caps.
//
// The same fixture is read by the web app's Jest suite, so both clients are held
// to one set of numbers. See packages/contracts/README.md.
//
// Each case runs through the real notifiers rather than recomputing the rule:
// a fixture checked against a second copy of the arithmetic would agree with
// itself while the app did something else.
import 'dart:convert';

import 'package:fabtrades/core/logic/free_limits.dart';
import 'package:fabtrades/core/models/subscription_status.dart';
import 'package:fabtrades/core/models/trade.dart';
import 'package:fabtrades/core/providers.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

import '../support/fixtures.dart';
import 'contract_fixtures.dart';

class _FreeSubscription extends SubscriptionNotifier {
  @override
  Future<SubscriptionStatus> build() async => SubscriptionStatus.free;
}

Trade _trade(String id) => Trade(id: id, createdAt: DateTime.now());

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  final contract = loadContract('free_limits');
  final limits = Map<String, dynamic>.from(contract['limits'] as Map);
  final cases = contractCases(contract, 'cases');

  /// A free account whose history already holds [trades].
  ///
  /// Seeded through storage rather than by calling `addTrade` in a loop, because
  /// that trims as it goes and could never produce an overfull history. Arriving
  /// overfull is a real state all the same: a lapsed subscription, or a sync that
  /// pulled down what another device saved while Pro.
  Future<ProviderContainer> freeContainer({int trades = 0}) async {
    SharedPreferences.setMockInitialValues({
      if (trades > 0)
        'saved_trades': jsonEncode([
          for (var i = 0; i < trades; i++) _trade('seeded-$i').toJson(),
        ]),
    });
    final prefs = await SharedPreferences.getInstance();
    final container = ProviderContainer(
      overrides: [
        paywallsRemovedProvider.overrideWithValue(false),
        sharedPreferencesProvider.overrideWithValue(prefs),
        subscriptionProvider.overrideWith(_FreeSubscription.new),
      ],
    );
    addTearDown(container.dispose);
    await container.read(subscriptionProvider.future);
    return container;
  }

  group('free limits contract', () {
    test('agrees with the shared caps', () {
      expect(FreeLimits.binderCards, limits['binderCards']);
      expect(FreeLimits.wantListCards, limits['wantListCards']);
      expect(FreeLimits.savedTrades, limits['savedTrades']);
      expect(FreeLimits.loanedCards, limits['loanedCards']);
    });

    for (final testCase in cases.where(
      (c) => c['limit'] == 'binderCards' || c['limit'] == 'wantListCards',
    )) {
      test(testCase['name'] as String, () async {
        final container = await freeContainer();
        final binder = container.read(binderProvider.notifier);
        final isWanted = testCase['limit'] == 'wantListCards';

        for (var i = 0; i < (testCase['existing'] as int); i++) {
          binder.add(buildCard(id: 'card-$i'), isWanted: isWanted);
        }

        expect(
          binder.add(buildCard(id: 'one-more'), isWanted: isWanted),
          testCase['allowed'],
        );
      });
    }

    for (final testCase in cases.where((c) => c['limit'] == 'loanedCards')) {
      test(testCase['name'] as String, () async {
        final container = await freeContainer();
        final lend = container.read(lendProvider.notifier);
        final groupId = lend.createGroup(isBorrowing: false);

        for (var i = 0; i < (testCase['existing'] as int); i++) {
          expect(lend.addCard(groupId, buildCard(id: 'lent-$i')), isTrue);
        }

        expect(
          lend.addCard(groupId, buildCard(id: 'one-more')),
          testCase['allowed'],
        );
      });
    }

    for (final testCase in cases.where((c) => c['limit'] == 'savedTrades')) {
      test(testCase['name'] as String, () async {
        final container = await freeContainer(
          trades: testCase['existing'] as int,
        );
        final history = container.read(tradeHistoryProvider.notifier);

        final trimmed = history.addTrade(_trade('newest'));

        // Saving is never refused, only trimmed: confirming a trade also
        // reconciles the binder, so blocking it would lose real information.
        expect(testCase['allowed'], true);
        expect(trimmed, testCase['trimmed']);

        final saved = container.read(tradeHistoryProvider);
        expect(saved.first.id, 'newest');
        expect(
          saved,
          hasLength(
            testCase['trimmed'] == 0
                ? (testCase['existing'] as int) + 1
                : FreeLimits.savedTrades,
          ),
        );
      });
    }
  });
}
